homeassistant/components/recorder/purge.py

```diff
@@ -7,32 +7,62 @@
 import homeassistant.util.dt as dt_util
 
 from .models import Events, RecorderRuns, States
-from .util import session_scope
+from .util import execute, session_scope
 
 _LOGGER = logging.getLogger(__name__)
 
 
-def purge_old_data(instance, purge_days, repack):
-    """Purge events and states older than purge_days ago."""
+def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
+    """Purge events and states older than purge_days ago.
+
+    Cleans up an timeframe of an hour, based on the oldest record.
+    """
     purge_before = dt_util.utcnow() - timedelta(days=purge_days)
     _LOGGER.debug("Purging events before %s", purge_before)
 
     try:
         with session_scope(session=instance.get_session()) as session:
-            deleted_rows = (
+            query = session.query(States).order_by(States.last_updated.asc()).limit(1)
+            states = execute(query, to_native=True, validate_entity_ids=False)
+
+            states_purge_before = purge_before
+            if states:
+                states_purge_before = min(
+                    purge_before, states[0].last_updated + timedelta(hours=1)
+                )
+
+            deleted_rows_states = (
                 session.query(States)
-                .filter(States.last_updated < purge_before)
+                .filter(States.last_updated < states_purge_before)
                 .delete(synchronize_session=False)
             )
-            _LOGGER.debug("Deleted %s states", deleted_rows)
+            _LOGGER.debug("Deleted %s states", deleted_rows_states)
 
-            deleted_rows = (
+            query = session.query(Events).order_by(Events.time_fired.asc()).limit(1)
+            events = execute(query, to_native=True)
+
+            events_purge_before = purge_before
+            if events:
+                events_purge_before = min(
+                    purge_before, events[0].time_fired + timedelta(hours=1)
+                )
```

Comment:
I am not 100% sure but I think the `states_purge_before` timestamp should be reused when deleting events.

With the timestamps unaligned, you might try to delete an event that is still referenced by the states table. That should fail due to the foreign key constraint.

Reply:
There is a relation defined, but it is not hard enforced. I've managed to get my states out of wack with the events tables during testing (like 60+ days difference) and it did not cause any issues.

If this was a concern, to begin with, the purge as before this change already worked by date & time, which didn't guarantee any referential integrity either.

Reply:
> There is a relation defined, but it is not hard enforced.

Hmm interesting. It seems like referential integrity must be enabled explicitly for SQLite and that we do not do that. I was not aware of this.

However, referential integrity is enabled by default for MySQL(InnoDB) and it also seems like we do not disable that so I think my concern is still valid.

> If this was a concern, to begin with, the purge as before this change already worked by date & time, which didn't guarantee any referential integrity either.

Integrity was implicitly guaranteed before by the invariant that `state_changed` events never fire before the state has changed.

Say we delete states and then events up to time `T`. Any stray state would have a time `> T`. However, because of that invariant, its matching event would also have time `> T` and it would thus not have been deleted. So the `State` was not stray after all.

With the new implementation we delete states up to `Ts` and events up to `Te`. They will be close but I cannot see that `Ts >= Te`.

Reply:
How is the foreign key defined in MySQL, Cascade delete?

Reply:
The default in MySQL is no action (which is restricting), fail is referential integrity is broken.

Reply:
Reading up a bit on SQLAlchemy cascades actually got me confused so I moved my dev setup to MySQL to test this a bit. After removing the state deletion from the code (to simulate the situation I am talking about) I got this error:
```
2020-06-29 09:21:28 WARNING (Recorder) [homeassistant.components.recorder.purge] Error purging history: (MySQLdb._exceptions.IntegrityError) (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`hadev`.`states`, CONSTRAINT `states_ibfk_1` FOREIGN KEY (`event_id`) REFERENCES `events` (`event_id`))')
[SQL: DELETE FROM events WHERE events.time_fired < %s]
[parameters: (datetime.datetime(2020, 6, 29, 7, 21, 28, 294027, tzinfo=<UTC>),)]
(Background on this error at: http://sqlalche.me/e/13/gkpj).
```

So it indeed seems like an issue if one is unfortunate enough to hit the small window.

@frenck If you want the easiest attack, just pick one of the timestamps and use that. Any timestamp should work, as long as the same one is used with both deletes.